hoodie-client/src/main/java/com/uber/hoodie/table/DefaultBulkInsertPartitioner.java

```diff
@@ -0,0 +1,187 @@
+ * - Output spark partition will have records from only one hoodie partition.
+ */
+public class DefaultBulkInsertPartitioner<T extends HoodieRecordPayload> extends Partitioner implements BulkInsertPartitioner<T> {
+    private static Logger logger = LogManager.getLogger(DefaultBulkInsertPartitioner.class);
+    private static final int BUCKET_MULTIPLIER = 4;
+    private int expectedOutputSparkPartitions;
+    private int numBuckets;
+    private HoodiePartitionMapper hoodiePartitionMapper;
+    private int numOfPartitions = -1;
+
+    private void init(JavaRDD<HoodieRecord<T>> records, int outputSparkPartitions) {
+        expectedOutputSparkPartitions = Math.max(1, outputSparkPartitions);
+        numBuckets = expectedOutputSparkPartitions * BUCKET_MULTIPLIER;
+        this.hoodiePartitionMapper = new HoodiePartitionMapper();
+        TreeMap<String, Long> bucketRecordCount = mapRecordsToBucket(records);
+        Counter totalRecordCount = new Counter(0);
+        // Count total records.
+        bucketRecordCount.values().forEach(v -> totalRecordCount.addCount(v));
+        long expectedRecordsPerSparkPartition = totalRecordCount.getCount() / expectedOutputSparkPartitions;
+        Map<String, Counter> currHoodiePartitionRecordCnt = new HashMap<>();
+        for (Entry<String, Long> entry : bucketRecordCount.entrySet()) {
+            String[] parts = entry.getKey().split(" ");
+            String hoodiePartition = parts[0];
+            if (!currHoodiePartitionRecordCnt.containsKey(hoodiePartition)) {
+                currHoodiePartitionRecordCnt.put(hoodiePartition, new Counter(0));
+                numOfPartitions++;
+            }
+            Counter cnt = currHoodiePartitionRecordCnt.get(hoodiePartition);
+            if (cnt.getCount() >= expectedRecordsPerSparkPartition) {
+                // Reset count and also change output spark partition. That way going forward records for same hoodie
+                // partition will be sent to different spark output partition.
+                cnt.setCount(0);
+                numOfPartitions++;
+            } else {
+                cnt.addCount(entry.getValue());
+            }
+            hoodiePartitionMapper.add(entry.getKey(), numOfPartitions);
+        }
+    }
+
+    @Override
+    public int getPartition(final Object key) {
+        Tuple2<HoodieRecord<T>, Integer> pair = (Tuple2<HoodieRecord<T>, Integer>) key;
+        return hoodiePartitionMapper.get(getBucketKey(pair._1(), pair._2()));
+    }
+
+    @Override
+    public int numPartitions() {
+        return numOfPartitions + 1;
+    }
+
+    public JavaRDD<HoodieRecord<T>> repartitionRecords(JavaRDD<HoodieRecord<T>> records, int outputSparkPartitions) {
+        init(records, outputSparkPartitions);
+        // As we are going over input records twice; we need to cache input records.
+        JavaRDD<HoodieRecord<T>> cachedRecords = records.persist(StorageLevel.MEMORY_AND_DISK_SER());
```

Comment:
RDDWrapper is actually just waiting for Hoodie to finish writing. This code is exactly same for both the runs.

Yes you are right. Overall it is very close to UpsertPartitioner but there are fundamental differences in it. It's better if we chat f2f. :) will ping you.